core/services/synchronization/stats_pusher.go

```diff
@@ -82,9 +85,11 @@ func (sp *StatsPusher) Close() error {
 		sp.cancel()
 	}
 	gormCallbacksMutex.Lock()
-	callbacks := sp.ORM.DB.Callback()
-	callbacks.Create().Remove(createCallbackName)
-	callbacks.Update().Remove(updateCallbackName)
+	_ = sp.ORM.DangerouslyUseRawDB(func(db *gorm.DB) error {
```

Comment:
Why is this `DangerouslyUseRawDB` and above it's `RawDB`?

Reply:
Forgot about this.  I added this while still unfamiliar with the ORM/Store code, because `StatsPusher` was being incorrectly torn down after `ORM`.  This meant that `StatsPusher` attempted to obtain an advisory lock after the lock had been released by `ORM` (in a non-reacquirable way) simply to deregister some GORM callbacks, an operation which doesn't touch the DB, therefore needing no lock.  However, I've fixed the order in which the store tears itself down, so this is no longer needed.

As for why the lock is not reacquirable by the node after being released, it's essentially because of the tradeoff that its semaphore semantics faces us with.  To release a Postgres advisory lock that has been acquired N times, you have two options:

1. Call `pg_advisory_unlock(x)` N times
2. Kill the connection to the DB (Postgres automatically releases the lock after a configurable delay)

I worry that option 1 will be either impossible or expensive, given that every call to the DB first has to acquire the lock, incrementing the semaphore each time.  A long-running node may need to issue many thousands of `pg_advisory_unlock(x)` calls (perhaps more, not sure) upon shutdown to properly release the lock so that the failover node can grab it.

Option 2 is clean and simple, especially given that the node only ever needs to unlock upon shutdown.  The DB connection will be killed anyway.  Moreover, this harmonizes the solution with the case where the node loses its connection by accident, rather than as part of an Unlock.

The semaphore semantics of the advisory lock seem to be intended more for cases where multiple clients are expected to have (mutexed) access to the same DB, but our case is really about exclusive access until failure/shutdown.  I'm led to wonder if there might be a better mechanism for this, but the current solution does seem to work for now.